### PR TITLE
SONAR-7453 Rename rules fields "effort_to_fix_description", "coefficient" and "offset

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/rule/RuleOperations.java [hunk 1]
```diff
@@ -108,13 +108,13 @@ public boolean updateRule(RuleDto ruleDto, @Nullable String newFunction,
...
-    return isSameRemediationFunction(newFunction, newCoefficient, newOffset, ruleDto.getDefaultRemediationFunction(), ruleDto.getDefaultRemediationCoefficient(),
-      ruleDto.getDefaultRemediationOffset());
+    return isSameRemediationFunction(newFunction, newCoefficient, newOffset, ruleDto.getDefaultRemediationFunction(), ruleDto.getDefaultRemediationGapMultiplier(),
+      ruleDto.getDefaultRemediationBaseEffort());
   }
 
   private static boolean isRuleDebtSameAsOverriddenValues(RuleDto ruleDto, @Nullable String newFunction,
     @Nullable String newCoefficient, @Nullable String newOffset) {
-    return isSameRemediationFunction(newFunction, newCoefficient, newOffset, ruleDto.getRemediationFunction(), ruleDto.getRemediationCoefficient(), ruleDto.getRemediationOffset());
+    return isSameRemediationFunction(newFunction, newCoefficient, newOffset, ruleDto.getRemediationFunction(), ruleDto.getRemediationGapMultiplier(), ruleDto.getRemediationBaseEffort());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RuleOperations.java]: MINOR Split this 186 characters long line (which is greater than 180 authorized). [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 2]
```diff
@@ -986,9 +994,18 @@ public DebtRemediationFunction debtRemediationFunction() {
       return debtRemediationFunction;
     }
 
+    /**
+     * @deprecated since 5.5, replaced by {@link #gapDescription()}
+     */
+    @Deprecated
     @CheckForNull
     public String effortToFixDescription() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java [hunk 3]
```diff
@@ -811,16 +811,24 @@ public NewRule setDebtRemediationFunction(@Nullable DebtRemediationFunction fn)
     }
 
     /**
+     * @deprecated since 5.5, replaced by {@link #setGapDescription(String)}
+     */
+    @Deprecated
+    public NewRule setEffortToFixDescription(@Nullable String s) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/rule/RulesDefinition.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java [hunk 4]
```diff
@@ -64,27 +64,42 @@ public Type type() {
 
   @Override
   @CheckForNull
+  @Deprecated
   public String coefficient() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)
MAJOR Add the missing @deprecated Javadoc tag. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java [hunk 5]
```diff
@@ -64,27 +64,42 @@ public Type type() {
...
+  @Override
+  @CheckForNull
+  public String gapMultiplier() {
+    return gapMultiplier;
   }
 
   @Override
   @CheckForNull
+  @Deprecated
   public String offset() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)
MAJOR Add the missing @deprecated Javadoc tag. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java [hunk 6]
```diff
@@ -63,35 +63,66 @@
...
+    Type(boolean usesGapMultiplier, boolean usesBaseEffort) {
+      this.usesGapMultiplier = usesGapMultiplier;
+      this.usesBaseEffort = usesBaseEffort;
     }
 
+    /**
+     * @deprecated since 5.5, replaced by {@link #usesGapMultiplier()}
+     */
+    @Deprecated
     public boolean usesCoefficient() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java [hunk 7]
```diff
@@ -63,35 +63,66 @@
...
 
   Type type();
 
   /**
-   * Non-null value on {@link Type#LINEAR} and {@link Type#LINEAR_OFFSET} functions, else {@code null}.
+   * @deprecated since 5.5, replaced by {@link #gapMultiplier()}
    */
+  @Deprecated
   @CheckForNull
   String coefficient();
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java [hunk 8]
```diff
@@ -63,35 +63,66 @@
...
+   */
+  @CheckForNull
+  String gapMultiplier();
+
+  /**
+   * @deprecated since 5.5, replaced by {@link #baseEffort()}
    */
+  @Deprecated
   @CheckForNull
   String offset();
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java [hunk 9]
```diff
@@ -63,35 +63,66 @@
...
+
+    public boolean usesGapMultiplier() {
+      return usesGapMultiplier;
     }
 
+    /**
+     * @deprecated since 5.5, replaced by {@link #usesBaseEffort()}
+     */
+    @Deprecated
     public boolean usesOffset() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/DebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 8 info

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues:
* MAJOR 2 major
* MINOR 1 minor
* INFO 8 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/rule/ws/RuleMapper.java [hunk 10]
```diff
@@ -281,51 +300,29 @@ private static boolean ruleToOverloaded(RuleDto rule) {
...
-          return ruleDto.getRemediationCoefficient();
-        }
-
-        @Override
-        public String offset() {
-          return ruleDto.getRemediationOffset();
-        }
-      };
+      return new DefaultDebtRemediationFunction(
+        DebtRemediationFunction.Type.valueOf(function.toUpperCase()),
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/rule/ws/RuleMapper.java]: MAJOR Define the locale to be used in this String operation. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1449)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java [hunk 11]
```diff
@@ -62,29 +62,50 @@ public Type type() {
     return type;
   }
 
+  /**
+   * @deprecated since 5.5, replaced by {@link #gapMultiplier}
+   */
   @Override
   @CheckForNull
+  @Deprecated
   public String coefficient() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java [hunk 12]
```diff
@@ -62,29 +62,50 @@ public Type type() {
...
+    return gapMultiplier;
+  }
+
+  /**
+   * @deprecated since 5.5, replaced by {@link #baseEffort}
+   */
   @Override
   @CheckForNull
+  @Deprecated
   public String offset() {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/debt/internal/DefaultDebtRemediationFunction.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues:
* MAJOR 1 major
* INFO 8 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/rule/ws/RuleMapper.java (hunk 10 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/rule/ws/RuleMapper.java]: don't forget to apply this feedback

Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/rule/ws/RuleMapper.java]: indeed

**Comment by teryk** [comment]
api/rules/update parameters should be updated (deprecate the existing one and add since 5.5 on the new parameters)

**Comment by teryk** [comment]
the rest seems good

**Pull request merged**